### PR TITLE
Avoid re-interpreting the rewrite lemma when it won't be cleared

### PR DESCRIPTION
**Pull request by SkySkimmer**

### CODE REVIEW

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by SkySkimmer** [comment]
@coqbot bench

File: tactics/equality.ml [hunk 1]
```diff
@@ -534,16 +534,21 @@ let general_rewrite_clause l2r with_evars ?tac c cl =
...
-    let sigma = Tacmach.project gl in
-    let env = Proofview.Goal.env gl in
-    try
-      let (sigma, (c, bl)) = f env sigma in
-      let c = try Some (destVar sigma c) with DestKO -> None in
-      apply_clear_request clear_flag (use_clear_hyp_by_default ()) c
-    with
-      e when noncritical e -> tclIDTAC
-  end
+  (* Ideally we would not need this shortcut to avoid uselessly runninf [f]
```
Comment (herbelin) [inline tactics/equality.ml:537]:
```suggestion
  (* Ideally we would not need this shortcut to avoid uselessly running [f]
```

**Comment by SkySkimmer** [comment]
Doesn't seem to have much impact, do we want to merge?

**Comment by ppedrot** [comment]
I think that this fast-path cannot hurt, but I'm not a huge fan of the way this PR is written. It's not robust to a change to `apply_clear_request`.

**Comment by SkySkimmer** [comment]
What change do you suggest to make it more robust?

**Comment by ppedrot** [comment]
I think we should have two variants of `apply_clear_request`, one taking a bool option (like the current one) and one taking just a bool. Furthermore, with a bit of GADT typing we may statically check whether the Id.t option argument is really needed, so that we wouldn't have this kind of missing matrix of checks going on.

**Comment by ppedrot** [comment]
It's worse than that: if I'm not mistaken, the API in Equality is only called with Some true and None. We should have a dedicated internal function rather than sharing weird intricacies of tactic internals.

**Comment by SkySkimmer** [comment]
Seems like nobody wants to merge this, let's move on

**Pull request closed without merging**